mesa/experimental/devs/eventlist.py

```diff
@@ -0,0 +1,191 @@
+from enum import IntEnum
+from heapq import heapify, heappop, heappush
+from types import MethodType
+from typing import Any, Callable
+from weakref import WeakMethod, ref
+
+
+class InstanceCounterMeta(type):
+    """Metaclass to make instance counter not share count with descendants
+
+    TODO:: can also be used for agents.unique_id
+    """
+
+    def __init__(cls, name, bases, attrs):
+        super().__init__(name, bases, attrs)
+        cls._ids = itertools.count(1)
+
+
+class Priority(IntEnum):
+    LOW = 10
+    DEFAULT = 5
+    HIGH = 1
+
+
+class SimulationEvent(metaclass=InstanceCounterMeta):
+    """A simulation event
+
+    the callable is wrapped using weakrefs, so there is no need to explicitly cancel event if e.g., an agent
+    is removed from the simulation.
+
+    Attributes:
+        time (float): The simulation time of the event
+        function  (Callable): The function to execute for this event
+        priority (Priority): The priority of the event
+        unique_id (int) the unique identifier of the event
+        function_args (list[Any]): Argument for the function
+        function_kwargs (Dict[str, Any]): Keyword arguments for the function
+
+    """
+
+    @property
+    def CANCELED(self) -> bool:
+        return self._canceled
+
+    def __init__(
+        self,
+        time: int | float,
+        function: Callable,
+        priority: Priority = Priority.DEFAULT,
+        function_args: list[Any] | None = None,
+        function_kwargs: dict[str, Any] | None = None,
+    ) -> None:
+        super().__init__()
+        self.time = time
+        self.priority = priority.value
+        self._canceled = False
+
+        if isinstance(function, MethodType):
+            function = WeakMethod(function)
+        else:
+            function = ref(function)
+
+        self.fn = function
+        self.unique_id = next(self._ids)
+        self.function_args = function_args if function_args else []
+        self.function_kwargs = function_kwargs if function_kwargs else {}
+
+        if self.fn is None:
+            raise Exception()
+
+    def execute(self):
+        """execute this event"""
+        fn = self.fn()
+        if fn is not None:
+            fn(*self.function_args, **self.function_kwargs)
+
+    def cancel(self) -> None:
+        """cancel this event"""
+        self._canceled = True
+        self.fn = None
+        self.function_args = []
+        self.function_kwargs = {}
+
+    def __cmp__(self, other):
+        if self.time < other.time:
+            return -1
+        if self.time > other.time:
+            return 1
+
+        if self.priority > other.priority:
+            return -1
+        if self.priority < other.priority:
+            return 1
+
+        if self.unique_id < other.unique_id:
+            return -1
+        if self.unique_id > other.unique_id:
+            return 1
+
+        # theoretically this should be impossible unless it is the
+        # exact same event
+        return 0
```

Comment:
Should an exception be raised instead? A duplicate event comparison seems to be a bug to me.

Reply:
I have been thinking about this code myself as well. This is one of the things I want to check when writing unit tests.

1. `__cmp__` should return -1, 0, or 1, so an exception should be raised in the heapq because that needs to produce a complete ordering
2. I think this code is fully redundant and probably not even called (to be tested) because the event lists consist of tuples (time, priority, unique_id, actual_event) which are sorted by heapq.

Reply:
I haven't used heapq before, so not an expert either, but from my understanding and reading this indeed is used by heapq for its sorting, so I do thing its needed and implemented correctly. I think even the 'return 0'  doesn't necessarily require the comment. Because this indeed just compares two events and it is a perfectly valid question to ask "are these 2 events the same event"

Reply:
Okay I should have read the whole implementation first. You are just adding the _to_tuple to the heap, so it might indeed never be called. Is this required for heap? Or could you not add just the event and then it uses the comparison?

Reply:
I have been otherwise occupied over the last few days. Hopefully, I have some time in the coming week to finally test this. Once testes, I'll report back here.